### PR TITLE
Bring back tombstones to Head block

### PR DESCRIPTION
**Pull request by codesome**
This PR reverts the changes done by https://github.com/prometheus/tsdb/pull/270 which is required for https://github.com/prometheus/prometheus/issues/6377.

One change that I noticed while fixing the tests is that even if all the samples are deleted in the tombstones, the querier would still return series but with 0 samples. Is this behaviour fine or should the querier not return any series at all?

### CODE REVIEW

**Comment by brian-brazil** [comment]
That behaviour is okay, we'd need a bit much pre-processing otherwise. PromQL should in principle already deal with this, but add uniitests if you're unsure.

**Comment by codesome** [comment]
Then the current test coverage looks fine for this part where samples are verified.

**Comment by brian-brazil** [comment]
What do benchmarks look like?

It also looks like once we delete data that you can't recreate it in that time range, which makes things simpler (and is reasonable semantically).

**Comment by codesome** [comment]
> What do benchmarks look like?

Query benchmarks? Haven't done any benchmark for this change yet

> It also looks like once we delete data that you can't recreate it in that time range

That was the behavior before prometheus/tsdb#270

**Comment by brian-brazil** [comment]
We don't have any benchmarks covering tombstones in the head?

**Comment by codesome** [comment]
I don't see any, so, nope. What do we want to benchmark for tombstones? Queries/iterators?

**Comment by brian-brazil** [comment]
It's not a use case we particularly care about performance wise, as you shouldn't be deleting from the head outside of rare mistakes you need to clean up so I wouldn't worry about it.

Probably should add a unittest to cover trying to add a sample in a time period that has been tombstones though.

**Comment by codesome** [comment]
@brian-brazil added the test cases.

Ping @krasi-georgiev for a review.

**Comment by brian-brazil** [comment]
:+1:

File: tsdb/head.go [hunk 1]
```diff
@@ -1088,39 +1068,8 @@ func (h *Head) Delete(mint, maxt int64, ms ...*labels.Matcher) error {
...
-
-	ms.reset()
-	for it.Next() {
-		t, v := it.At()
-		ok, _ := ms.append(t, v)
-		if !ok {
-			level.Warn(h.logger).Log("msg", "failed to add sample during delete")
-		}
+	for _, s := range stones {
+		h.tombstones.AddInterval(s.Ref, s.Intervals[0])
```
Comment (bwplotka) [inline tsdb/head.go]: Can we do that directly in `for p.Next() {` loop above?

Comment (codesome) [reply tsdb/head.go]: There are cases where things would error out above, in which case we don't want to add any stones for the delete call. So adding here assures that there was no error during the delete process above.

File: tsdb/head.go [hunk 2]
```diff
@@ -1068,13 +1054,7 @@ func (h *Head) Delete(mint, maxt int64, ms ...*labels.Matcher) error {
...
-			stones = append(stones, tombstones.Stone{Ref: p.At(), Intervals: tombstones.Intervals{{Mint: t0, Maxt: t1}}})
-		}
-		if err := h.chunkRewrite(p.At(), tombstones.Intervals{{Mint: t0, Maxt: t1}}); err != nil {
-			return errors.Wrap(err, "delete samples")
-		}
-		dirty = true
+		stones = append(stones, tombstones.Stone{Ref: p.At(), Intervals: tombstones.Intervals{{Mint: t0, Maxt: t1}}})
 	}
 	if p.Err() != nil {
 		return p.Err()
```
[the remark this answers is not in the document]
Comment (codesome) [reply tsdb/head.go]: Thanks, removed it.

File: tsdb/head.go (hunk 1 above)
Comment (bwplotka) [reply tsdb/head.go]: :+1:  Thanks for explanation!

**Pull request merged**